src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      stateCopy = addProperties(stateCopy, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      stateCopy = removeProperties(stateCopy, action.keysToRemove);
+    } else if (action.type === 'clear') {
+      stateCopy = clear(stateCopy);
+    }
+
+    stateCopies.push({ ...stateCopy });
```

Comment:
To adhere to the 'Don't Repeat Yourself' principle, the action of cloning the state copy (line 22) should be performed once after the switch or if-else block, not in every case.